floodlight/models/kinematics.py

```diff
@@ -0,0 +1,102 @@
+import warnings
+from dataclasses import dataclass
+
+import numpy as np
+
+from floodlight.utils.types import Numeric
+from floodlight import Pitch, XY
+from floodlight.core.property import PlayerProperty
+
+
+@dataclass
+class ModelDistance:
+    """Class for calculating distances of players on the pitch.
+
+    Parameters
+    __________
+    pitch: Pitch
+        Pitch object of the data. Should at least include information about the data's
+        unit. If pitch.unit is 'percent', length and width of the pitch are also needed.
+
+    """
+
+    pitch: Pitch = None
+    length: Numeric = None
+    width: Numeric = None
+    unit: str = None
+
+    def __init__(self, pitch):
+        self.unit = pitch.unit
+        if pitch.length is not None:
+            self.length = pitch.length
+        if pitch.width is not None:
+            self.width = pitch.width
+        if pitch.unit is not None:
+            self.unit = pitch.unit
+        else:
+            warnings.warn(
+                "No information about pitch unit. Coordinates are assumed to be in m!"
+            )
+        self._distance_euclidean = None
+        self.framerate = None
+
+    def __str__(self):
+        return (
+            f"Floodlight ModelDistance Class with unit [{self.unit}],"
+            f"length {self.length} and width {self.width}"
+        )
+
+    def fit(self, xy: XY, direction: str = None):
+        """Fits a model to calculate euclidean distances to an xy object.
+
+        Parameters
+        ----------
+        xy: XY
+            Floodlight XY Data object.
+        direction: str
+            One of {'x', 'y', 'xy'} of None.
+            Indicates the direction to calculate the distance.
+            'x' will only calculate distance in the direction of the x-coordinate.
+            'y' will only calculate distance in the direction of the y-coordinate.
+            'xy' of None will calculate distance in both directions.
+
+        """
+
+        distance_euclidean = None
+
+        if self.unit is None or self.unit == "m":
+            unit_factor = (1, 1)
+        elif self.unit == "cm":
+            unit_factor = (100, 100)
+        elif self.unit == "percent":
+            unit_factor = (self.length / 100, self.width / 100)
+        else:
+            warnings.warn(
+                "No information about pitch unit. Coordinates are assumed to be in m!"
+            )
+            unit_factor = (1, 1)
```

Comment:
as above

Reply:
I'm throwing a ValueError here because it is possible to instantiate a Pitch object with any kind of attribute for unit. Not sure if this is the right place for it though. Shouldn't the Pitch object check if its attributes are valid?

Reply:
That's a good question which comes down to a design choice in general. We've chosen to follow the "better ask for forgiveness" path here, let me try to clarify that:
If you would enforce certain values of ``pitch.unit`` during instantiation, this would restrict it's usage throughout the package and also require to decide on all possible values it may take (which I doubt we know at this point). This design produces an undesirable bi-directional coupling: On the one hand, you would have to think about all possible functions that could work with the pitch object when deciding on the legal values of ``pitch.unit``. On the other hand, during implementation of e.g. this functionality, you need to think of all possible values for ``pitch.unit``, which couples your function to a different object. If a new value was to be added later on (e.g. 'yards'), this might break your (unrelated) code here.

To avoid this, the alternative is: ``pitch`` needs to have an ``unit`` attribute, that's the minimal requirement. But it doesn't matter what that is. If the user wants to have it in 'yards', so be it. Your function here might not work then, but others will (e.g. plotting it). On the other hand, you do not need to worry about _all_ legal values here, but rather decide what your function needs: Maybe everything that is metric (m, cm, dm, mm, ...) is fine. So by querying here and raising an Error if your requirements are not satisfied, you're politely saying "do whatever you want, but if you want to use this method, your data better be in ``['m', 'cm', 'percent', ...]``

Reply:
Makes total sense